libyul/Object.cpp

```diff
@@ -129,33 +141,60 @@ std::set<YulString> Object::qualifiedDataNames() const
 	return qualifiedNames;
 }
 
-std::vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
+void Object::visitPath(YulString _qualifiedName, std::function<bool(Object const*)> const& _visitor) const
 {
-	yulAssert(_qualifiedName != name, "");
-	yulAssert(subIndexByName.count(name) == 0, "");
-
 	if (boost::algorithm::starts_with(_qualifiedName.str(), name.str() + "."))
 		_qualifiedName = YulString{_qualifiedName.str().substr(name.str().length() + 1)};
 	yulAssert(!_qualifiedName.empty(), "");
 
 	std::vector<std::string> subObjectPathComponents;
 	boost::algorithm::split(subObjectPathComponents, _qualifiedName.str(), boost::is_any_of("."));
 
-	std::vector<size_t> path;
 	Object const* object = this;
 	for (std::string const& currentSubObjectName: subObjectPathComponents)
 	{
 		yulAssert(!currentSubObjectName.empty(), "");
-		auto subIndexIt = object->subIndexByName.find(YulString{currentSubObjectName});
-		yulAssert(
-			subIndexIt != object->subIndexByName.end(),
-			"Assembly object <" + _qualifiedName.str() + "> not found or does not contain code."
-		);
-		object = dynamic_cast<Object const*>(object->subObjects[subIndexIt->second].get());
-		yulAssert(object, "Assembly object <" + _qualifiedName.str() + "> not found or does not contain code.");
-		yulAssert(object->subId != std::numeric_limits<size_t>::max(), "");
-		path.push_back({object->subId});
+
+		YulString subObjectName = YulString{currentSubObjectName};
+		object = object->at(subObjectName);
```

Comment:
Just a general remark about naming: please avoid renaming things in a way that does not make it clear what the distinction between them really is. Here both `subObjectName` and `currentSubObjectName` are pretty much the same thing and the difference is not that one is not "current". It's that one is a `YulString`. So something like `currentSubObjectNameYul` would be a much clearer name for the former. But in this situation I find that `YulString{currentSubObjectName}` is not much longer so I'd just use it directly:

```suggestion
		object = object->at(YulString{currentSubObjectName});
```